bpo-42972: Track sqlite3 statement objects

erlend-aasland opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-42972](https://bugs.python.org/issue42972) -->
https://bugs.python.org/issue42972
<!-- /issue-number -->

Commit:
bpo-42972: Track sqlite3 statement objects

pablogsal commented on Modules/_sqlite/cursor.c:
```diff
@@ -512,6 +512,7 @@ _pysqlite_query_execute(pysqlite_Cursor* self, int multiple, PyObject* operation
             goto error;
         }
         rc = pysqlite_statement_create(self->statement, self->connection, operation);
```
Tracking should be done on the constructor, not on the code calling it, as is not it's responsability. This way there is no need for synchronization and eliminates the possibility of getting a badly crated object.

erlend-aasland replied:
Yes, that's better. I'll move the allocation there as well, for consistency.

erlend-aasland replied:
I've moved allocation and GC tracking over to `pysqlite_statement_create`. PTAL.

Commit:
Allocate and track statement objects in pysqlite_statement_create

By allocating and tracking creation of statement object in
pysqlite_statement_create(), the caller does not need to worry about GC
syncronization, and eliminates the possibility of getting a badly
created object. All related fault handling is moved to
pysqlite_statement_create().

Commit:
Remove now unused constants

vstinner commented on Modules/_sqlite/statement.c:
```diff
@@ -48,29 +48,37 @@ typedef enum {
+    self->st = NULL;
+    self->in_use = 0;
     self->in_weakreflist = NULL;
     self->sql = Py_NewRef(sql);
```
Could you initialize is_dml and db here?

erlend-aasland replied:
Sure! I tried to minimise the changes, but nothing would make me more glad than to refactor this a little bit more :)

erlend-aasland replied:
May I add a commit for intermingled variable declaration?

vstinner commented on Modules/_sqlite/statement.c:
```diff
@@ -48,29 +48,37 @@ typedef enum {
-        rc = PYSQLITE_SQL_WRONG_TYPE;
-        return rc;
+        PyErr_SetString(pysqlite_Warning,
+                        "SQL is of wrong type. Must be string.");
```
You may add Py_TYPE(sql)->tp_name in the error message.

```suggestion
                        "SQL is of wrong type ('%s'). Must be string.");
```

erlend-aasland replied:
BTW; this should probably have been a `TypeError`, not a `sqlite3.Warning`. Changing this is probably far out of scope for this PR.

vstinner reviewed (approved):
LGTM, but I would like to see the minor issues being addressed first before merging your change.

Commit:
Improve error message

Co-authored-by: Victor Stinner <vstinner@python.org>

Commit:
Group and order member initialisation

vstinner commented on Modules/_sqlite/statement.c:
```diff
@@ -48,35 +48,45 @@ typedef enum {
     self->sql = Py_NewRef(sql);
+    self->in_use = 0;
+    self->is_dml = 0;
+    self->in_weakreflist = NULL;
```
Thanks, it looks better like that. It makes me me confident that calling Py_DECREF(self) is safe in this function.

erlend-aasland replied:
Yes, this is way better. Thanks for pushing the PR in this direction.

vstinner reviewed (approved):
LGTM.
